src/main/scala/io/iohk/ethereum/domain/Blockchain.scala

```diff
@@ -326,6 +336,13 @@ class BlockchainImpl(
       stateRootHash,
       noEmptyAccounts = false
     )
+
+  def nodesKeyValueStorageFor(blockNumber: Option[BigInt]): NodesKeyValueStorage =
+    PruningMode.nodesKeyValueStorage(pruningMode, nodeStorage)(blockNumber)
+
+  def pruneState(blockNumber: BigInt): Unit = PruningMode.prune(pruningMode, blockNumber, nodeStorage)
```

Comment:
Currently, we are using `pruneState` only during block save in regularSync, shouldn't we also prune our state trie during fast sync?

Reply:
No, because while doing fast sync you are downloading the whole state trie for a single block, the target block. Therefore there is no need to prune anything.

Reply:
As I understand currently `PruningMode.prune(pruningMode, blockNumber, nodeStorage)` only delete stale snapshots from db.

so If during fastsync we are saving nodes using https://github.com/input-output-hk/mantis/blob/b022f9657fa8b76f9352992de2a3935763b7685f/src/main/scala/io/iohk/ethereum/domain/Blockchain.scala#L269

and we would not use `prune` anywhere,  wouldn't we create and save unnecessary snapshots for whole chain? 

I just did a little test (10000 blocks) of fastsync on this branch and `phase/deadalus`, and size of `phase/deadalus` datadir is 53,7 MB, and with new pruning it is 61,5 MB, there is a little increase.

Reply:
We will run `fastSyncTarget` pruning when doing regular sync on the block `fastSyncTarget`+ `history`. So, for example if blockchain head is block `4745467`: 
- we will pick `4744967` as `fastSyncTarget` (checkout `target-block-offset` config, default to 500)
- When executing regular sync (as it's after `fastSyncTarget`) on block `4746467` (`history` is 1000 by default) we will prune `4744967`

I think the size increase might be because we are storing the Snapshots (which contains the whole node) instead of the previous buggy version (where we were only storing  block numbers to delete when we detected zero references)